src/Analyzers/CSharp/Analyzers/UsePrimaryConstructor/CSharpUsePrimaryConstructorDiagnosticAnalyzer.cs

```diff
@@ -196,204 +199,237 @@ private void OnSymbolEnd(SymbolAnalysisContext context)
-                if (!styleOption.Value)
-                    return;
+                        // Look to see if we have trivial `_x = x` or `this.x = x` assignments.  If so, then the field/prop
+                        // could be a candidate for removal (as long as we determine that all use sites of the field/prop would
+                        // be able to use the captured primary constructor parameter).
 
-                // only classes/structs can have primary constructors (not interfaces, enums or delegates).
-                if (namedType.TypeKind is not (TypeKind.Class or TypeKind.Struct))
-                    return;
+                        if (analyzer._candidateMembersToRemove.Count > 0)
+                            context.RegisterOperationAction(analyzer.AnalyzeFieldOrPropertyReference, OperationKind.FieldReference, OperationKind.PropertyReference);
 
-                // Don't want to offer on records.  It's completely fine for records to not use primary constructs and
-                // instead use init-properties.
-                if (namedType.IsRecord)
-                    return;
+                        if (namedType == context.Symbol)
+                        {
+                            context.RegisterSymbolEndAction(analyzer.OnSymbolEnd);
+                            hasAnalyzerForContextSymbol = true;
+                        }
+                    }
 
-                // No need to offer this if the type already has a primary constructor.
-                if (namedType.TryGetPrimaryConstructor(out _))
-                    return;
+                    namedType = namedType.ContainingType;
+                }
 
-                // Need to see if there is a single constructor that either calls `base(...)` or has no constructor
-                // initializer (and thus implicitly calls `base()`), and that all other constructors call `this(...)`.
-                if (!TryFindPrimaryConstructorCandidate(out var primaryConstructor, out var primaryConstructorDeclaration))
-                    return;
+                if (hasAnyAnalyzer && !hasAnalyzerForContextSymbol)
+                {
+                    // Workaround for https://github.com/dotnet/roslyn/issues/68484
+                    // If require analysis for any containing type, but not the context's named type,
+                    // we register a dummy SymbolEndAction for context's named type.
+                    context.RegisterSymbolEndAction(_ => { });
+                }
 
-                if (primaryConstructor.Parameters.Length == 0)
-                    return;
+                return;
 
-                if (primaryConstructor.DeclaredAccessibility != Accessibility.Public)
-                    return;
+                Analyzer? TryGetOrCreateAnalyzer(INamedTypeSymbol namedType)
+                    => stateMap.GetOrAdd(namedType, TryCreateAnalyzer);
```

Comment:
i don't love this, as it will great a mapping from *every* named type in the world in this map (even if it points to null).